clients/drcachesim/reader/reader.cpp

```diff
@@ -67,231 +67,251 @@ reader_t::operator++()
+        if (bundle_idx_ == input_entry_->size)
+            bundle_idx_ = 0;
+        break;
+    case TRACE_TYPE_INSTR_FLUSH:
+    case TRACE_TYPE_DATA_FLUSH:
+        assert(cur_tid_ != 0 && cur_pid_ != 0);
+        cur_ref_.flush.pid = cur_pid_;
+        cur_ref_.flush.tid = cur_tid_;
+        cur_ref_.flush.type = (trace_type_t)input_entry_->type;
+        cur_ref_.flush.size = input_entry_->size;
+        cur_ref_.flush.addr = input_entry_->addr;
+        if (cur_ref_.flush.size != 0)
+            have_memref = true;
+        break;
+    case TRACE_TYPE_INSTR_FLUSH_END:
+    case TRACE_TYPE_DATA_FLUSH_END:
+        cur_ref_.flush.size = input_entry_->addr - cur_ref_.flush.addr;
+        have_memref = true;
+        break;
+    case TRACE_TYPE_THREAD:
+        cur_tid_ = (memref_tid_t)input_entry_->addr;
+        // tid2pid might not be filled in yet: if so, we expect a
+        // TRACE_TYPE_PID entry right after this one, and later asserts
+        // will complain if it wasn't there.
+        cur_pid_ = tid2pid_[cur_tid_];
+        break;
+    case TRACE_TYPE_THREAD_EXIT:
+        cur_tid_ = (memref_tid_t)input_entry_->addr;
+        cur_pid_ = tid2pid_[cur_tid_];
+        assert(cur_tid_ != 0 && cur_pid_ != 0);
+        // We do pass this to the caller but only some fields are valid:
+        cur_ref_.exit.pid = cur_pid_;
+        cur_ref_.exit.tid = cur_tid_;
+        cur_ref_.exit.type = (trace_type_t)input_entry_->type;
+        have_memref = true;
+        break;
+    case TRACE_TYPE_PID:
+        cur_pid_ = (memref_pid_t)input_entry_->addr;
+        // We do want to replace, in case of tid reuse.
+        tid2pid_[cur_tid_] = cur_pid_;
+        break;
+    case TRACE_TYPE_MARKER:
+        cur_ref_.marker.type = (trace_type_t)input_entry_->type;
+        if (!online_ &&
+            (input_entry_->size == TRACE_MARKER_TYPE_VERSION ||
+             input_entry_->size == TRACE_MARKER_TYPE_FILETYPE)) {
+            // Do not carry over a prior thread on a thread switch to a
+            // first-time-seen new thread, whose tid entry is *after* these
+            // markers for offline traces.
+            cur_pid_ = 0;
+            cur_tid_ = 0;
+        } else {
+            assert(cur_tid_ != 0 && cur_pid_ != 0);
         }
+        cur_ref_.marker.pid = cur_pid_;
+        cur_ref_.marker.tid = cur_tid_;
+        cur_ref_.marker.marker_type = (trace_marker_type_t)input_entry_->size;
+        cur_ref_.marker.marker_value = input_entry_->addr;
+        // Look for timestamp+cpu duplicated from the prior chunk.  Skip them on
+        // a linear walk.  File readers that support seeking will read them
+        // and use them to start post-seek iteration.
+        if (chunk_instr_count_ > 0 &&
+            cur_ref_.marker.marker_type == TRACE_MARKER_TYPE_TIMESTAMP &&
+            cur_instr_count_ / chunk_instr_count_ !=
+                last_timestamp_instr_count_ / chunk_instr_count_) {
```

Comment:
This assumes that each chunk always has a duplicated timestamp+cpu from the previous chunk, which is fine for second-and-onwards chunks. About the first chunk though, we'll have just one timestamp+cpu, which we don't want to skip. I see last_timestamp_instr_count_ is init to 0, so looks like this will indeed skip the (only) timestamp+cpu in the first chunk. Am I missing something? Does this WAI?

Reply:
Since last_timestamp_instr_count_ is init to 0, cur_instr_count_/chunk_instr_count_ will be 0 which will match last_timestamp_instr_count_ / chunk_instr_count_ as also 0 and so we will not skip, which is what we want.

Reply:
Ah oops. Got it.